src/tdigest.c

```diff
@@ -0,0 +1,512 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+
+/*
+ * Interface for t-digest support
+ *
+ * Originally from PipelineDB 1.0.0
+ * Original copyright (c) 2018, PipelineDB, Inc. and released under Apache License 2.0
+ * Modifications copyright by Timescale, Inc. per NOTICE
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdlib.h>
+
+#include "postgres.h"
+#include "tdigest.h"
+#include "utils/elog.h"
+#include "utils/memutils.h"
+#include "utils/palloc.h"
+
+/* scale function */
+#define integrated_location(compression, q) ((compression) * (asin(2 * (q) -1) + M_PI / 2) / M_PI)
+
+#define interpolate(x, x0, x1) (((x) - (x0)) / ((x1) - (x0)))
+#define float_eq(f1, f2) (fabs((f1) - (f2)) <= FLT_EPSILON)
+
+typedef struct mergeArgs
+{
+	TDigest *t;
+	Centroid *centroids;
+	int index;
+	float8 weight_so_far;
+	float8 k1;
+	float8 min;
+	float8 max;
+} mergeArgs;
+
+/*
+ * Based on compression level, estimates ideal number of centroids to store in input buffer
+ * before merging into main centroid storage
+ */
+static uint32
+estimate_compression_threshold(int compression)
+{
+	compression = Min(TDIGEST_MAX_COMPRESSION, Max(TDIGEST_MIN_COMPRESSION, compression));
+	return (uint32)(7.5 + 0.37 * compression - 2e-4 * compression * compression);
+}
+
+TDigest *
+ts_tdigest_create(void)
+{
+	return ts_tdigest_create_with_compression(TDIGEST_DEFAULT_COMPRESSION);
+}
+
+TDigest *
+ts_tdigest_create_with_compression(int compression)
+{
+	uint32 size;
+	TDigest *t;
+
+	if (compression < TDIGEST_MIN_COMPRESSION || compression > TDIGEST_MAX_COMPRESSION)
+		elog(ERROR,
+			 "compression parameter should be in [%d, %d], got %d",
+			 TDIGEST_MIN_COMPRESSION,
+			 TDIGEST_MAX_COMPRESSION,
+			 compression);
+
+	size = ceil(compression * M_PI / 2) + 1;
+
+	/* palloc0 means no manual nil'ing/setting to 0 of values necessary */
+	t = palloc0(sizeof(TDigest) + size * sizeof(Centroid));
+
+	t->compression = 1.0 * compression;
+	t->threshold = estimate_compression_threshold(compression);
+	t->size = size;
+	t->min = INFINITY;
+
+	SET_VARSIZE(t, ts_tdigest_size(t));
```

Comment:
I looked at the code for tdigest_recv and send. Looks like we are assuming we can have int64 bytes alloced for a t-digest. Instead we need to add some memory checks to the T-digest alloc code (@jlockerman are there any gucs /functions used in PG to limit memory usage by a process that can be used here). This will avoid 32-bit porting and OOM issues.

Reply:
v1_len can be used to store the alloced bytes and then reused by tdigest_send/recv instead of recomputing the size. I don't see v1_len actually being used. Is it?